### PR TITLE
Chrono fix

### PR DESCRIPTION
**Pull request by antmat**
subj

### CODE REVIEW

File: chrono/include/cocaine/chrono.hpp [hunk 1]
```diff
@@ -69,6 +69,7 @@ class chrono_t:
         streamed<io::timer_id_t>
         set_timer_impl(double first, double repeat, bool send_id);
 
+        static constexpr uint64_t minimal_interval_us = 10000;
```
Comment (3Hren) [inline chrono/include/cocaine/chrono.hpp]: Why not to use `std::chrono`/`boost::chrono` types instead?

File: chrono/src/chrono.cpp [hunk 2]
```diff
@@ -53,6 +53,10 @@ chrono_t::notify_every(double time, bool send_id) {
 cocaine::streamed<io::timer_id_t>
 chrono_t::set_timer_impl(double first, double repeat, bool send_id) {
     streamed<io::timer_id_t> promise;
+    if(repeat != 0.0 && repeat * 1000000.0 < minimal_interval_us) {
```
Comment (3Hren) [inline chrono/src/chrono.cpp]: Comparing with `0.0` is not specified. Comparing with zero duration is specified - use it.

File: chrono/src/chrono.cpp [hunk 3]
```diff
@@ -70,7 +74,9 @@ chrono_t::set_timer_impl(double first, double repeat, bool send_id) {
     try {
         ptr->insert(std::make_pair(timer_id, desc));
 
-        desc.timer_->expires_from_now(boost::posix_time::seconds(first));
+        // We get time as double seconds, so convert it to integer microseconds.
+        boost::int64_t microsecond_time = static_cast<boost::int64_t>(first * 1000000.0);
```
Comment (3Hren) [inline chrono/src/chrono.cpp]: Use standard types instead.

File: chrono/src/chrono.cpp [hunk 4]
```diff
@@ -99,7 +105,9 @@ void
 chrono_t::restart(io::timer_id_t timer_id) {
     timer_desc_t& timer_desc = timers_.unsafe().at(timer_id);
 
-    timer_desc.timer_->expires_from_now(boost::posix_time::seconds(timer_desc.interval_));
+    // We get time as double seconds, so convert it to integer microseconds.
+    boost::int64_t microsecond_time = static_cast<boost::int64_t>(timer_desc.interval_ * 1000000.0);
```
Comment (3Hren) [inline chrono/src/chrono.cpp]: \ditto

File: chrono/src/chrono.cpp [hunk 5]
```diff
@@ -128,17 +136,23 @@ chrono_t::on_timer(const std::error_code& ec, io::timer_id_t timer_id) {
             restart(timer_id);
         }
     } catch (std::exception& ex) {
-        COCAINE_LOG_ERROR(log_, "possibly bug in timer service in chrono_t::on_timer(), error is %s", ex.what());
+        COCAINE_LOG_WARNING(log_, "timer was canceled during on_timer, exception: %s", ex.what());
```
Comment (3Hren) [inline chrono/src/chrono.cpp]: So it's now not a bug?

File: chrono/src/chrono.cpp [hunk 6]
```diff
@@ -128,17 +136,23 @@ chrono_t::on_timer(const std::error_code& ec, io::timer_id_t timer_id) {
...
-        timers_.unsafe().at(timer_id).promise_.close();
-    } catch (const std::exception& ex) {
-        COCAINE_LOG_ERROR(log_, "error occured while removing timer %s", ex.what());
+    // Already locked.
+    auto it = timers_.unsafe().find(timer_id);
+    if(it != timers_.unsafe().end()) {
+        try {
+            it->second.promise_.close();
+        } catch (const std::exception& ex) {
+            COCAINE_LOG_ERROR(log_, "could not remove timer - %s", ex.what());
```
Comment (3Hren) [inline chrono/src/chrono.cpp]: We closing the channel - not removing timers.

File: chrono/src/chrono.cpp (hunk 5 above)
Comment (antmat) [inline chrono/src/chrono.cpp]: no

File: chrono/src/chrono.cpp (hunk 3 above)
Comment (antmat) [inline chrono/src/chrono.cpp]: what's the difference? boost::posix_time::microseconds accepts boost::int64_t

File: chrono/src/chrono.cpp (hunk 2 above)
Comment (antmat) [inline chrono/src/chrono.cpp]: it's double comparision. I don't understand what do you mean.

File: chrono/include/cocaine/chrono.hpp (hunk 1 above)
Comment (antmat) [inline chrono/include/cocaine/chrono.hpp]: it's not constexpr. Besides there is a lot of things to improve (we are using doubles for times).

Comment (3Hren) [inline chrono/include/cocaine/chrono.hpp]: It **is** constexpr.

File: chrono/src/chrono.cpp (hunk 2 above)
Comment (3Hren) [inline chrono/src/chrono.cpp]: I mean that the line `if (repeat < min_interval) { ...` where time point types are strongly typed reads better than this.

**Comment by antmat** [comment]
I sticked down to a minimal fix. If we plan to use it in future - we need to rework it

**Pull request merged**